cmd/sonobuoy/app/images.go

```diff
@@ -0,0 +1,293 @@
+	if _, err := os.Stat(imagesflags.e2eRegistryConfig); os.IsNotExist(err) {
+		errlog.LogError(errors.Errorf("file does not exist or cannot be opened: %v", imagesflags.e2eRegistryConfig))
+		os.Exit(1)
+	}
+
+	// Check if username if specified for Registry Auth
+	if len(imagesflags.username) > 0 {
+		fmt.Print("Registry password: ")
+		bytePassword, err := terminal.ReadPassword(int(syscall.Stdin))
+		if err != nil {
+			errlog.LogError(errors.Wrap(err, "couldn't get password from user"))
+			os.Exit(1)
+		}
+
+		imagesflags.password = string(bytePassword)
+		fmt.Print("\n")
+	}
+
+	switch imagesflags.plugin {
+	case "e2e":
+
+		cfg, err := imagesflags.kubeconfig.Get()
+		if err != nil {
+			errlog.LogError(errors.Wrap(err, "couldn't get REST client"))
+			os.Exit(1)
+		}
+
+		sbc, err := getSonobuoyClient(cfg)
+		if err != nil {
+			errlog.LogError(errors.Wrap(err, "could not create sonobuoy client"))
+			os.Exit(1)
+		}
+
+		version, err := sbc.Version()
+		if err != nil {
+			errlog.LogError(errors.Wrap(err, "couldn't get Sonobuoy client"))
+			os.Exit(1)
+		}
+
+		// Get list of upstream images that match the version
+		upstreamRegistry, err := image.NewRegistryList("", version)
+		if err != nil {
+			errlog.LogError(errors.Wrap(err, "couldn't init Registry List"))
+			os.Exit(1)
+		}
+
+		upstreamImages, err := upstreamRegistry.GetImageConfigs()
+		if err != nil {
+			errlog.LogError(errors.Wrap(err, "couldn't get images for version"))
+			os.Exit(1)
+		}
+
+		// Get list of images that match the version
+		privateRegistry, err := image.NewRegistryList(imagesflags.e2eRegistryConfig, version)
+		if err != nil {
+			errlog.LogError(errors.Wrap(err, "couldn't init Registry List"))
+			os.Exit(1)
+		}
+
+		privateImages, err := privateRegistry.GetImageConfigs()
+		if err != nil {
+			errlog.LogError(errors.Wrap(err, "couldn't get images for version"))
+			os.Exit(1)
+		}
+
+		ctx := context.Background()
+		cli, err := client.NewClientWithOpts(client.WithVersion("1.37"))
+		if err != nil {
+			panic(err)
+		}
+
+		for k, v := range upstreamImages {
+			downloadImage(ctx, cli, v)
+			tagImage(ctx, cli, v, privateImages[k])
+			pushImage(ctx, cli, privateImages[k])
+			fmt.Println("########")
+		}
+
+	default:
+		errlog.LogError(errors.Errorf("Unsupported plugin: %v", imagesflags.plugin))
+		os.Exit(1)
+	}
+}
+
+func downloadImage(ctx context.Context, cli *client.Client, img image.Config) {
+	out, err := cli.ImagePull(ctx, img.GetE2EImage(), types.ImagePullOptions{})
+	if err != nil {
+		errlog.LogError(errors.Wrapf(err, "error downloading image: %v", img.GetE2EImage()))
+	}
+	defer out.Close()
+
+	// Show status
+	err = streamDockerMessages(out)
+	if err != nil {
+		errlog.LogError(errors.Wrapf(err, "error downloading image: %v", img.GetE2EImage()))
+	}
+	// Wait for all data to complete
+	_, err = ioutil.ReadAll(out)
+	if err != nil {
+		errlog.LogError(errors.Wrapf(err, "error downloading image: %v", img.GetE2EImage()))
```

Comment:
This pattern is used a few times in the code; could you elaborate a bit on it?

An error on this `ioutil.ReadAll` would indicate the read failed, not the image failed to download. It seems like we'd be doing ourselves a disservice to err the download (or tag or push) if this read fails. And if we don't even want the data that we are reading, why are we doing the read at all?

Reply:
The read is reading the file stream. If this fails then the download failed. This blocks to wait for the image to finish, otherwise, they all go in parallel and we don't want 20 images downloading concurrently.